### PR TITLE
[Merged by Bors] - feat: expand API for `Polynomial.divX` and some related results

### PR DESCRIPTION
**Pull request by BoltonBailey**
Various lemmas about `Polynomial.natDegree`

Co-authored-by: Pratyush Mishra

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/Polynomial/Degree/Definitions.lean [hunk 1]
```diff
@@ -1132,6 +1132,21 @@ theorem eq_C_of_natDegree_eq_zero (h : natDegree p = 0) : p = C (coeff p 0) :=
 theorem eq_C_coeff_zero_iff_natDegree_eq_zero : p = C (p.coeff 0) ↔ p.natDegree = 0 :=
   ⟨fun h ↦ by rw [h, natDegree_C], eq_C_of_natDegree_eq_zero⟩
 
+theorem eq_C_one_of_Monic_natDegree_zero (hf : p.Monic) (hfd : p.natDegree = 0) : p = C 1 := by
```
Comment (alexjbest) [inline Mathlib/Data/Polynomial/Degree/Definitions.lean:1135]:
```suggestion
theorem eq_C_one_of_monic_natDegree_zero (hf : p.Monic) (hfd : p.natDegree = 0) : p = C 1 := by
```
perhaps? I'm not 100% sure of the naming convention but it seems strange that you have two different styles next to each other

File: Mathlib/Data/Polynomial/Div.lean [hunk 2]
```diff
@@ -219,12 +219,21 @@ theorem degree_modByMonic_le (p : R[X]) {q : R[X]} (hq : Monic q) : degree (p %
...
+
 end Ring
 
 section CommRing
 
 variable [CommRing R] {p q : R[X]}
 
+theorem X_dvd_sub_C : X ∣ p - C (p.coeff 0) := by
+  rw [X_dvd_iff]
+  rw [coeff_sub, coeff_C]
```
Comment (alexjbest) [inline Mathlib/Data/Polynomial/Div.lean:234]:
```suggestion
  rw [X_dvd_iff, coeff_sub, coeff_C]
```
or even just put it in one big simp

**Comment by adomani** [comment]
I think that adding some API around `divX` will help here.  For instance, this is an alternative proof of `natDegree_divX_eq_natDegree_tsub_one` that uses neither `Nontriviality` nor `NoZeroDivisors`:
```
import Mathlib.Data.Polynomial.EraseLead
import Mathlib.Data.Polynomial.Inductions

open Polynomial

-- no NoZeroDivisors assumption
variable {R : Type u} [Semiring R] {p : R[X]} {a : R}

namespace Polynomial

@[simp]
theorem divX_zero : divX (0 : R[X]) = 0 := leadingCoeff_eq_zero.mp rfl

@[simp]
theorem divX_one : divX (1 : R[X]) = 0 := by
  ext
  simpa only [coeff_divX, coeff_zero] using coeff_one (_ + 1)

theorem divX_C_mul : divX (C a * p) = C a * divX p := by
  ext
  simp

theorem divX_X_pow : divX (X ^ n : R[X]) = if (n = 0) then 0 else X ^ (n - 1) := by
  cases n
  · simp
  · ext n
    simp [coeff_X_pow]

noncomputable
def divX_hom : R[X] →+ R[X] :=
{ toFun := divX
  map_zero' := divX_zero
  map_add' := fun _ _ => divX_add }

@[simp] theorem divX_hom_toFun : divX_hom p = divX p := rfl

theorem natDegree_divX_eq_natDegree_tsub_one : p.divX_hom.natDegree = p.natDegree - 1 := by
  nontriviality R  -- gets rid of the nontriviality assumption
  apply map_natDegree_eq_sub (φ := divX_hom)
  · intro f
    simpa [divX_hom, divX_eq_zero_iff] using eq_C_of_natDegree_eq_zero
  · intros n c c0
    rw [← C_mul_X_pow_eq_monomial, divX_hom_toFun, divX_C_mul, divX_X_pow]
    split_ifs with n0 <;> simp [n0]
    exact natDegree_C_mul_X_pow (n - 1) c c0

theorem natDegree_divX_le : p.divX.natDegree ≤ p.natDegree :=
(natDegree_divX_eq_natDegree_tsub_one).trans_le (Nat.pred_le _)
```

**Comment by BoltonBailey** [comment]
@adomani Do you have any opinion on whether `divX` should exist at all? All these lemmas seem great but I feel like I might want them for `p /ₘ X` as well, so why even have this function?

File: Mathlib/Data/Polynomial/Inductions.lean [hunk 3]
```diff
@@ -51,6 +52,20 @@ theorem divX_mul_X_add (p : R[X]) : divX p * X + C (p.coeff 0) = p :=
...
+theorem coeff_zero_eq_iff {a : R} : p.coeff 0 = a ↔ ∃ g : Polynomial R, p = X * g + C a := by
+  constructor
+  · intro h
+    use p.divX
+    rw [←h, X_mul_divX_add p]
+  · intro h
+    cases' h with g hg
+    rw [hg, add_comm, coeff_add, coeff_C]
+    simp only [eq_self_iff_true, if_true, mul_coeff_zero, coeff_X_zero, MulZeroClass.zero_mul,
+      add_zero]
```
Comment (adomani) [inline Mathlib/Data/Polynomial/Inductions.lean:67]:
```suggestion
theorem coeff_zero_eq_iff {a : R} : p.coeff 0 = a ↔ ∃ g : Polynomial R, p = X * g + C a := by
  conv_rhs => rw [← X_mul_divX_add p]
  exact ⟨fun h => ⟨p.divX, by congr⟩, fun ⟨g, h⟩ => by simpa using congr_arg (fun f => coeff f 0) h⟩
```
This is a shorter version, but I wonder if you really need this lemma.

File: Mathlib/Data/Polynomial/Degree/Definitions.lean [hunk 4]
```diff
@@ -1132,6 +1132,15 @@ theorem eq_C_of_natDegree_eq_zero (h : natDegree p = 0) : p = C (coeff p 0) :=
 theorem eq_C_coeff_zero_iff_natDegree_eq_zero : p = C (p.coeff 0) ↔ p.natDegree = 0 :=
   ⟨fun h ↦ by rw [h, natDegree_C], eq_C_of_natDegree_eq_zero⟩
 
+theorem eq_C_one_of_monic_natDegree_zero (hf : p.Monic) (hfd : p.natDegree = 0) : p = C 1 := by
```
Comment (jcommelin) [inline Mathlib/Data/Polynomial/Degree/Definitions.lean:1135]: Shouldn't this be `p = 1`?
```suggestion
theorem eq_one_of_monic_natDegree_zero (hf : p.Monic) (hfd : p.natDegree = 0) : p = 1 := by
```

Comment (BoltonBailey) [reply Mathlib/Data/Polynomial/Degree/Definitions.lean]: Good point

File: Mathlib/Data/Polynomial/Inductions.lean [hunk 5]
```diff
@@ -51,6 +52,9 @@ theorem divX_mul_X_add (p : R[X]) : divX p * X + C (p.coeff 0) = p :=
 set_option linter.uppercaseLean3 false in
 #align polynomial.div_X_mul_X_add Polynomial.divX_mul_X_add
 
+theorem X_mul_divX_add (p : R[X]) : X * divX p + C (p.coeff 0) = p :=
```
[the remark this answers is not in the document]
Comment (ocfnash) [reply Mathlib/Data/Polynomial/Inductions.lean:55]: You can also add `simp` to `divX_mul_X_add` above. Of course neither will fire very often but there's not harm having them.

**Review by ocfnash — APPROVED** [review]
Please apply the suggestions and then feel free to merge

bors d+

**Pull request closed without merging**